### DISCUSSION TITLE
Show reality data in Viewer component

### DISCUSSION THREAD

**Discussion by ovaisa** [iTwin/itwinjs-core#1443] [General] [answered]
I created web viewer app using as outlined in the Getting Started >> Online (Quick Start) >> 3. Get, build, and develop code

     https://www.itwinjs.org/learning/tutorials/develop-web-viewer/

Now added code in the app to view reality data (following the sample in frontend-sample-showcase>> Reality Data)

const availableModels: ContextRealityModelProps[] = await findAvailableUnattachedRealityModels(iModelConnection.contextId!, iModelConnection);

It doesn;t seem to get the info on reality model in the project.  I am using the contextID and modelID of the imodel in iModelHub.  It has both CAD and reality data.  When I view the model in iModel Manager, I set both CAD and reality.  The viewer app only displays CAD data (because the reality info is not returned).

Ovais Ashraf

**Comment by pmconne** [comment, marked as answer]
@williamkbentley it appears applications registered via the dashboard lack the "reality-data-:read" scope?

**Reply by ovaisa** [reply]
Correct. The reality-data-:read directive is not in scope by default.  GlenW told me to add this keyword in the .env file (though his keyword didn't have '-' after 'data').  I tried both keywords (with/without dash) without success.  

The imodel was loaded by another company with ContextCapture account and I was added as participant to the Connected project.  Could it be that there is ContextCapture access issue (as I don't have licenses/account).  Though I can see both CAD and reality data in iModelManager in Connect portal.

**Reply by calebmshafer** [reply]
@ovaisa, the correct scope is `reality-data:read`, without the `-` so what you have should be working if GlenW helped you with it.

Can you look at the Network tab and see if there are any failed requests? If it is a permissions error, it should show up there.

**Reply by ovaisa** [reply]
Hi Caleb,

Thank you for your support.  I am attaching two files.  One contains the request headers and response when the findAvailableUnattached...() function is executed.  Second is the HAR file (renamed to .log).

Thank you for looking into this matter.


[Header+Response.txt](https://github.com/imodeljs/imodeljs/files/6509722/Header%2BResponse.txt)
[localhost.log](https://github.com/imodeljs/imodeljs/files/6509726/localhost.log)

**Reply by calebmshafer** [reply]
thanks @ovaisa. Based on the logs, it looks like you're getting a valid response from the Reality Data services. Specifically for this request in the `Header_Response.txt`, `https://connect-realitydataservices.bentley.com/v2.5/Repositories/S3MXECPlugin--d20678d8-ad8c-4732-a32f-5ec62e2befc0/S3MX/RealityData?project=d20678d8-ad8c-4732-a32f-5ec62e2befc0&$filter=Type+eq+%27RealityMesh3DTiles%27+or+Type+eq+%27OPC%27+or+Type+eq+%27OMR%27`.

Is the array `availableModels`, from your code, empty or just have the wrong data in it for the project?

**Reply by ovaisa** [reply]
availableModel is empty (length = 0)

Here is the code I added to app.tsx (copied from frontend-sample-showcase >> Reality Model

=====================
  const _oniModelReady = async (iModelConnection: IModelConnection) => {
    if (!iModelConnection) {
      return;
    }
    try {
      const availableModels: ContextRealityModelProps[] = await findAvailableUnattachedRealityModels(iModelConnection.contextId!, iModelConnection);
      for (const crmProp of availableModels) {
/// TODO - attach reality data to viewport style....
      }
    } catch (e) {
      // if failed, close the imodel and reset the state
      alert(e.message);
    }
 }
============================


From Watch window 

Block
crmProp: undefined
Block
availableModels: Array(0)
length: 0
__proto__: Array(0)
Local
iModelConnection: RemoteBriefcaseConnection
backgroundMapLocation: BackgroundMapLocation {_ecefValidated: false, _ecefToDb: Transform}
codeSpecs: CodeSpecs {_iModel: RemoteBriefcaseConnection}
displayedExtents: Range3d {low: Point3d, high: Point3d}
elements: Elements {_iModel: RemoteBriefcaseConnection}
geoServices: GeoServices {_iModel: RemoteBriefcaseConnection}
hilited: HiliteSet {iModel: RemoteBriefcaseConnection, _elements: HilitedElementIds, subcategories: HilitedIds, models: HilitedIds}
models: Models {_iModel: RemoteBriefcaseConnection, loaded: Map(0)}
name: "Norco_Test"
onClose: BeEvent {_listeners: Array(0), _insideRaiseEvent: false}
openMode: 1
rootSubject: {name: "Norco_Test"}
routingContext: IModelRoutingContext {token: RpcRoutingToken}
selectionSet: SelectionSet {iModel: RemoteBriefcaseConnection, _elements: Set(0), onChanged: BeEvent}
subcategories: SubCategoriesCache {_byCategoryId: Map(0), _appearances: Map(0), _imodel: RemoteBriefcaseConnection}
tiles: Tiles {_treesBySupplier: Map(0), _disposed: false, _iModel: RemoteBriefcaseConnection}
transientIds: TransientIdSequence {_localId: 0}
views: Views {_iModel: RemoteBriefcaseConnection}
_changeSetId: "9fa45ba5fa0caadd6514cea400dd8f66773a5ac9"
_contextId: "d20678d8-ad8c-4732-a32f-5ec62e2befc0"
_fileKey: "136d6576-73df-40c2-9f92-5c4d59184441:9fa45ba5fa0caadd6514cea400dd8f66773a5ac9"
_gcsDisabled: false
_globalOrigin: Point3d {x: 0, y: 0, z: 0}
_iModelId: "136d6576-73df-40c2-9f92-5c4d59184441"
_projectExtents: Range3d {low: Point3d, high: Point3d}
_reopenConnectionHandler: async (request, response, resubmit, reject) => {…}
_snapRpc: OneAtATimeAction {msg: "abandoned", _run: ƒ}
_toolTipRpc: OneAtATimeAction {msg: "abandoned", _run: ƒ}
changeSetId: (...)
contextId: (...)
ecefLocation: (...)
geographicCoordinateSystem: (...)
globalOrigin: (...)
iModelId: (...)
isBlank: (...)
isBriefcase: (...)
isClosed: (...)
isGeoLocated: (...)
isOpen: (...)
isReadonly: (...)
isSnapshot: (...)
key: (...)
noGcsDefined: (...)
projectExtents: (...)
__proto__: CheckpointConnection
this: undefined
Closure (App)
Closure (./src/App.tsx)
Window
Global

**Reply by pmconne** [reply]
@ovaisa do you get the same results if you use `findAvailableRealityModels` instead of `findAvailableUnattachedRealityModels`?
BTW, both of those functions are `@internal` - consider using [queryRealityData](https://github.com/imodeljs/imodeljs/blob/085bb9226da8612b3475767339e03a6213a16a0a/core/frontend/src/ContextRealityModelState.ts#L141) instead.

**Reply by ovaisa** [reply]
const availableModels: ContextRealityModelProps[] = await findAvailableRealityModels(iModelConnection.contextId!);

is returning 18 models (tilesets)...  

availableModels: Array(18)
0: {tilesetUrl: "https://connect-realitydataservices.bentley.com/v2…/RealityData/f6c272bf-6d33-43cc-a369-77b3529549d2", name: "Production_1", description: "RDS_Description", realityDataId: "f6c272bf-6d33-43cc-a369-77b3529549d2", orbitGtBlob: undefined}
1: {tilesetUrl: "https://connect-realitydataservices.bentley.com/v2…/RealityData/d7a8d697-df95-4551-84ed-2900d6d612cb", name: "Production_1-1", description: "RDS_Description", realityDataId: "d7a8d697-df95-4551-84ed-2900d6d612cb", orbitGtBlob: undefined}
2: {tilesetUrl: "https://connect-realitydataservices.bentley.com/v2…/RealityData/3f55ebda-f1d2-4010-88ea-07220bb1e302", name: "Production_1-2", description: "RDS_Description", realityDataId: "3f55ebda-f1d2-4010-88ea-07220bb1e302", orbitGtBlob: undefined}
3: {tilesetUrl: "https://connect-realitydataservices.bentley.com/v2…/RealityData/448b7477-39cf-4bf7-869a-fe4ecbe2bef9", name: "Production_1-3", description: "RDS_Description", realityDataId: "448b7477-39cf-4bf7-869a-fe4ecbe2bef9", orbitGtBlob: undefined}
4: {tilesetUrl: "https://connect-realitydataservices.bentley.com/v2…/RealityData/8c3fc6d6-5e20-4bd6-834d-257d2e6a11ff", name: "Production_1-4", description: "RDS_Description", realityDataId: "8c3fc6d6-5e20-4bd6-834d-257d2e6a11ff", orbitGtBlob: undefined}
5: {tilesetUrl: "https://connect-realitydataservices.bentley.com/v2…/RealityData/6761064e-8931-4923-a4c7-5a213e29ec83", name: "Production_1-5", description: "RDS_Description", realityDataId: "6761064e-8931-4923-a4c7-5a213e29ec83", orbitGtBlob: undefined}
6: {tilesetUrl: "https://connect-realitydataservices.bentley.com/v2…/RealityData/8145cfea-5e61-4445-9fcd-f808140784fa", name: "Production_1-6", description: "RDS_Description", realityDataId: "8145cfea-5e61-4445-9fcd-f808140784fa", orbitGtBlob: undefined}
7: {tilesetUrl: "https://connect-realitydataservices.bentley.com/v2…/RealityData/e03ceb9b-391b-4113-802f-2e4ddd4e1ff5", name: "Production_1-7", description: "RDS_Description", realityDataId: "e03ceb9b-391b-4113-802f-2e4ddd4e1ff5", orbitGtBlob: undefined}
8: {tilesetUrl: "https://connect-realitydataservices.bentley.com/v2…/RealityData/cf0f49eb-9150-4548-b0a2-49d5fe4f1f8a", name: "Production_1-8", description: "RDS_Description", realityDataId: "cf0f49eb-9150-4548-b0a2-49d5fe4f1f8a", orbitGtBlob: undefined}
9: {tilesetUrl: "https://connect-realitydataservices.bentley.com/v2…/RealityData/f1259040-8c38-4373-8ab8-aa4b726a9985", name: "Production_1-9", description: "RDS_Description", realityDataId: "f1259040-8c38-4373-8ab8-aa4b726a9985", orbitGtBlob: undefined}
10: {tilesetUrl: "https://connect-realitydataservices.bentley.com/v2…/RealityData/4c7ff3d5-7c19-4655-8b22-6e5a6cb02b27", name: "Production_1-10", description: "RDS_Description", realityDataId: "4c7ff3d5-7c19-4655-8b22-6e5a6cb02b27", orbitGtBlob: undefined}
11: {tilesetUrl: "https://connect-realitydataservices.bentley.com/v2…/RealityData/a52f0528-8e56-4660-a936-cd4bfb9bedc6", name: "Production_1-11", description: "RDS_Description", realityDataId: "a52f0528-8e56-4660-a936-cd4bfb9bedc6", orbitGtBlob: undefined}
12: {tilesetUrl: "https://connect-realitydataservices.bentley.com/v2…/RealityData/57398726-0901-4714-8f3e-08acc9bccd13", name: "Production_1-12", description: "RDS_Description", realityDataId: "57398726-0901-4714-8f3e-08acc9bccd13", orbitGtBlob: undefined}
13: {tilesetUrl: "https://connect-realitydataservices.bentley.com/v2…/RealityData/301ae344-502d-4fb8-ae67-7fe64f3759b7", name: "Production_1-13", description: "RDS_Description", realityDataId: "301ae344-502d-4fb8-ae67-7fe64f3759b7", orbitGtBlob: undefined}
14: {tilesetUrl: "https://connect-realitydataservices.bentley.com/v2…/RealityData/6a75bbda-a27e-4b88-9ad9-f115c829faff", name: "Production_1-14", description: "RDS_Description", realityDataId: "6a75bbda-a27e-4b88-9ad9-f115c829faff", orbitGtBlob: undefined}
15: {tilesetUrl: "https://connect-realitydataservices.bentley.com/v2…/RealityData/9ba2eae0-4dc7-4f85-975a-e611606f80ce", name: "Production_1-15", description: "RDS_Description", realityDataId: "9ba2eae0-4dc7-4f85-975a-e611606f80ce", orbitGtBlob: undefined}
16: {tilesetUrl: "https://connect-realitydataservices.bentley.com/v2…/RealityData/f4c1de59-56fd-445f-bdde-f5b1d7d0a2e2", name: "Overview_L.png", description: "RDS_Description", realityDataId: "f4c1de59-56fd-445f-bdde-f5b1d7d0a2e2", orbitGtBlob: undefined}
17: {tilesetUrl: "https://connect-realitydataservices.bentley.com/v2…/RealityData/e2ea25cc-9fc7-4ba2-bd6b-40c4ef528e39", name: "Close_Up.png", description: "RDS_Description", realityDataId: "e2ea25cc-9fc7-4ba2-bd6b-40c4ef528e39", orbitGtBlob: undefined}
length: 18

**Reply by pmconne** [reply]
That suggests that all 18 of those reality models are associated with geometric models stored in your iModel. Can you confirm that? There will be a `tilesetUrl` property in the models' JsonProperties fields.

**Reply by ovaisa** [reply]
So I removed the second argument of 

findAvailableUnattachedRealityModels(iModelConnection.contextId!); //, iModelConnection);

and it too is now returning 18 tilesets.  

Is there a sample code that shows how to use queryRealityData().  I am just following the code in sample sets so would be helpful if you can point to sample that is using this public (rather than internal) function.

Thank you

**Reply by ovaisa** [reply]
The reality model was created by my client, I will have to check with them.  I believe due to the size of the model as it was processed by ContextCapture in multiple (18?) tiles. 

If this is true, do I have to attach all these models one by one to display style or is there a single url that points to the entire set.

**Reply by pmconne** [reply]
If a geometric model exists in the iModel with a tilesetUrl pointing to a reality model, all you have to do is add that model's Id to the view's model selector. If there are multiple such models and you want to see all of their reality models, add all of their Ids to the model selector.

We are working on making sure the samples use only `public` APIs. If you follow my link to `queryRealityData` you can see exactly how to call it by looking at the two functions above it. But if the reality models are already persisted in the iModel you should have no need to query the reality data service anyway.

**Comment by ovaisa** [comment]
queryRealityData() is returning tilesetURLs!  Thank you Paul and Caleb for your assistance in resolving this matter.
